main.go

```diff
@@ -0,0 +1,377 @@
+		if err != nil {
+			exit(1, "error during creating output: %v", err)
+		}
+		defer f.Close()
+		printers.Printers = append(printers.Printers, &CsvPrinter{
+			Out: f,
+		})
+	}
+
+	printers.doPrint(func(printer Printer) {
+		printer.PrintHeader()
+	})
+
+	var dataFuncProvider func(cuCount *uint64) func(mtd *desc.MethodDescriptor, callData *runner.CallData) []byte
+
+	fmt.Println(options.Mode)
+	if options.Mode == "spam" {
+		var profile map[string]int64
+		if options.SpamProfile == "" {
+			profile = DEFAULT_PROFILE
+		} else {
+			profile = map[string]int64{}
+			profileRaw, _ := os.ReadFile(options.SpamProfile)
+			err := yaml.Unmarshal(profileRaw, profile)
+			if err != nil {
+				exit(1, "error during fetching spam profile: %v", err)
+			}
+		}
+		fmt.Println(profile)
+		dataFuncProvider = func(cuCount *uint64) func(mtd *desc.MethodDescriptor, callData *runner.CallData) []byte {
+			return NewEthSpamBinaryDataFunc(profile, options.SourceHost, dshackle.ChainRef(options.Chain), cuCount, context.Background())
+		}
+	} else if options.Mode == "prepared" {
+		files, err := os.ReadDir(options.PreparedRequests)
+		if err != nil {
+			exit(1, "error during fetching requests: %v", err)
+		}
+		reqs := make([]*dshackle.NativeCallRequest, 0)
+		for _, file := range files {
+			if file.IsDir() {
+				continue
+			}
+			name := options.PreparedRequests + "/" + file.Name()
+			raw, err := os.ReadFile(name)
+			if err != nil {
+				panic(err)
+			}
+			jsonrpc := JsonRpcRequest{}
+			err = json.Unmarshal(raw, &jsonrpc)
+			fmt.Printf("loading  %s\n", name)
+			if err != nil {
+				panic(err)
+			}
+			params, _ := json.Marshal(jsonrpc.Params)
+			req := &dshackle.NativeCallRequest{
+				Chain: dshackle.ChainRef(options.Chain),
+				Items: []*dshackle.NativeCallItem{
+					{
+						Id:      uint32(jsonrpc.Id),
+						Method:  jsonrpc.Method,
+						Payload: params,
+					},
+				},
+			}
+
+			if options.RequestLabel != "" {
+				selector := dshackle.Selector{}
+				_ = protojson.Unmarshal([]byte(options.RequestLabel), &selector)
+				req.Selector = &selector
+			}
+
+			fmt.Printf("request added to load: %s from %s\n", jsonrpc.Method, name)
+			reqs = append(reqs, req)
+		}
+		dataFuncProvider = func(cuCount *uint64) func(mtd *desc.MethodDescriptor, callData *runner.CallData) []byte {
+			pointer := 0
+			return func(mtd *desc.MethodDescriptor, callData *runner.CallData) []byte {
+				req := reqs[pointer]
+				pointer = (pointer + 1) % len(reqs)
+				data, _ := proto.Marshal(req)
+				if options.PreparedCU > 0 {
+					*cuCount += options.PreparedCU
+				} else {
+					*cuCount += gas.CountGas(req.Items[0].Method)
+				}
+				return data
+			}
+		}
+	} else {
+		exit(1, "unknown mode")
+	}
+
+	for _, l := range load {
+
+		printers.doPrint(func(printer Printer) {
+			printer.PrintPreLine(l)
+		})
+
+		var curCu uint64 = 0
+
+		ghzOpts := make([]runner.Option, 0)
+
+		ghzOpts = append(ghzOpts,
+			runner.WithBinaryDataFunc(dataFuncProvider(&curCu)),
```

Comment:
Looks like runner can call `dataFuncProvider` in parallel, this lead to update `curCu` concurrently. Better to replace it with atomic.